### PR TITLE
extmod: Add ubinascii.unhexlify

### PR DESCRIPTION
**Pull request by dhylands**
I also tweaked hexlify since its suppsoed to only take 1 arg. I'm not sure why it was coded to take 1 or 2 and ignore the 2nd.

### CODE REVIEW

**Comment by pfalcon** [comment]
> I also tweaked hexlify since its suppsoed to only take 1 arg. I'm not sure why it was coded to take 1 or 2 and ignore the 2nd.

Because I intend to add 2nd arg of separator between byte values, as an extension to CPython. I didn't do that so far because, as I mentioned, I put any non-standard additions on hold until more standard stuff is done, and we catch up on documentation. Taking 2nd arg is still a useful reminder of the upcoming functionality.

**Comment by dhylands** [comment]
> I also tweaked hexlify since its suppsoed to only take 1 arg. I'm not sure why it was coded to take 1 or 2 and ignore the 2nd.
> 
> Because I intend to add 2nd arg of separator between byte values, as an extension to CPython. I didn't do that so far because, as I mentioned, I put any non-standard additions on hold until more standard stuff is done, and we catch up on documentation. Taking 2nd arg is still a useful reminder of the upcoming functionality.

OK - restored hexlify to its original prototype and added a comment to explain why the second argument was intended for.

**Comment by pfalcon** [comment]
Thanks!

File: extmod/modubinascii.c [hunk 1]
```diff
@@ -58,10 +62,43 @@ STATIC mp_obj_t mod_binascii_hexlify(mp_uint_t n_args, const mp_obj_t *args) {
 }
 MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_binascii_hexlify_obj, 1, 2, mod_binascii_hexlify);
 
+STATIC mp_obj_t mod_binascii_unhexlify(mp_obj_t data) {
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(data, &bufinfo, MP_BUFFER_READ);
+
+    if ((bufinfo.len & 1) != 0) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_ubinascii_dot_Error, "Odd-length string"));
```
Comment (pfalcon) [inline extmod/modubinascii.c]: Sorry, this is obvious overkill. Raising ValueError is enough.

Comment (dhylands) [inline extmod/modubinascii.c]: Yeah - I wasn't sure how close to binascii we should be. I'm happy either way. I figured I'd throw this out and see what responses I got :)

**Comment by dhylands** [comment]
Replaced ubinascii.Error with ValueError

**Comment by pfalcon** [comment]
+1. There're some ways to save few more bytes there, but should be good enough for v1. Will merge tomorrow if @dpgeorge won't beat me.

**Comment by dhylands** [comment]
I did a couple iterations to get the size down. It started out being 0xC8 bytes then 0x9C bytes and finally 0x98 bytes (on stmhal).

I can see making it smaller if you make the assumption that the input is valid (i.e. it would convert some non-valid inputs).

If you didn't care to distinguish the types of invalid input, then you could save a bunch of bytes by only having one error message (i.e. "Invalid input" or something).

I'd be curious to hear any other suggestions.

**Comment by pfalcon** [comment]
> I'd be curious to hear any other suggestions.

Try using unichar_isxdigit(). Hex digit parsing is also done elsewhere, so would make sense either reuse it, or refactor to be reusable.

**Comment by dhylands** [comment]
Paradoxically, changing:
```C
if (hex_ch >= '0' && hex_ch <= '9') {
```
to be
```C
if (unichar_isdigit(hex_ch)) {
```
drops the size by 4 bytes, and changing 
```C
        } else if ((hex_ch >= 'A' && hex_ch <= 'F') ||
                   (hex_ch >= 'a' && hex_ch <= 'f')) {
```
to be
```C
        } else if (unichar_isxdigit(hex_ch)) {
```
drops the size by 4 bytes, but if I make both changes then I see no change in size.

If I expose the hex_digit function in py/lexer.c and call it:
```C
        if (unichar_isxdigit(hex_ch)) {
            hex_byte += hex_digit(hex_ch);
        } else {
```
then I save 20 bytes (function drops from 0x98 to 0x84)

**Comment by dhylands** [comment]
So I'd propose that we pull hex_digit from lexer.c and put it in unicode.c and call it unichar_hex_digit.

I also noticed that the unichar_isxdigit can be improved to use the lookup table (we can add a FL_XDIGIT). This will give a performance boost and code size reduction.

**Comment by dhylands** [comment]
Refactored hex_digit from py/lexer.c, reducing function size by 20 bytes (on stmhal)

File: extmod/modubinascii.c [hunk 2]
```diff
@@ -58,10 +60,39 @@ STATIC mp_obj_t mod_binascii_hexlify(mp_uint_t n_args, const mp_obj_t *args) {
...
+    vstr_t vstr;
+    vstr_init_len(&vstr, bufinfo.len / 2);
+    byte *in = bufinfo.buf, *out = (byte*)vstr.buf;
+    byte hex_byte = 0;
+    for (mp_uint_t i = bufinfo.len; i--;) {
+        byte hex_ch = *in++;
+        if (unichar_isxdigit(hex_ch)) {
+            hex_byte += unichar_hex_digit(hex_ch);
+        } else {
+            nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_ValueError, "Non-hexadecimal digit found"));
```
Comment (dpgeorge) [inline extmod/modubinascii.c]: "non-hex digit found" is shorter :)

File: extmod/modubinascii.c [hunk 3]
```diff
@@ -58,10 +60,39 @@ STATIC mp_obj_t mod_binascii_hexlify(mp_uint_t n_args, const mp_obj_t *args) {
 }
 MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_binascii_hexlify_obj, 1, 2, mod_binascii_hexlify);
 
+STATIC mp_obj_t mod_binascii_unhexlify(mp_obj_t data) {
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(data, &bufinfo, MP_BUFFER_READ);
+
+    if ((bufinfo.len & 1) != 0) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_ValueError, "Odd-length string"));
```
Comment (dpgeorge) [inline extmod/modubinascii.c]: Please start msg strings with lower case for consistency with others.

File: py/misc.h [hunk 4]
```diff
@@ -127,6 +127,7 @@ bool unichar_isupper(unichar c);
 bool unichar_islower(unichar c);
 unichar unichar_tolower(unichar c);
 unichar unichar_toupper(unichar c);
+mp_uint_t unichar_hex_digit(unichar c);
```
Comment (dpgeorge) [inline py/misc.h]: Better to call it unichar_xdigit_value.

**Comment by dpgeorge** [comment]
Thanks @dhylands.  My changes mentioned above are made and merged in 3ad94d6072555955a796c8c72a526c4fdff86711.

File: extmod/modubinascii.c (hunk 3 above)
Comment (dhylands) [inline extmod/modubinascii.c]: ok - I had originally copied the errors from CPython.

**Comment by dhylands** [comment]
For anybody not following along over on the forum, now that we have unhexlify, I've released rshell: https://github.com/dhylands/upy-shell/tree/master/rshell

**Pull request closed without merging**